contracts/Controller.sol

```diff
@@ -322,13 +376,58 @@ contract Controller is ReentrancyGuard, Ownable {
      * @dev Only account owner or operator can withdraw long option from vault
      * @param _args WithdrawArgs structure
      */
-    // function _withdrawLong(Actions.WithdrawArgs memory _args) internal isAuthorized(_args.owner) {}
+    function _withdrawLong(Actions.WithdrawArgs memory _args)
+        internal
+        isAuthorized(msg.sender, _args.owner)
+        isValidVaultId(_args.owner, _args.vaultId)
+        returns (MarginAccount.Vault memory)
+    {
+        OtokenInterface otoken = OtokenInterface(_args.asset);
+
+        require(now <= otoken.expiryTimestamp(), "Controller: can not withdraw an expired otoken");
+
+        vaults[_args.owner][_args.vaultId]._removeLong(address(otoken), _args.amount, _args.index);
+
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        marginPool.transferToUser(address(otoken), _args.to, _args.amount);
+
+        emit LongOtokenWithdrawed(address(otoken), _args.owner, _args.to, _args.vaultId, _args.amount);
+
+        return vaults[_args.owner][_args.vaultId];
+    }
 
     /**
      * @notice deposit collateral asset into vault
      * @param _args DepositArgs structure
      */
-    // function _depositCollateral(Actions.DepositArgs memory _args) internal {}
+    function _depositCollateral(Actions.DepositArgs memory _args)
+        internal
+        isValidVaultId(_args.owner, _args.vaultId)
+        returns (MarginAccount.Vault memory)
+    {
+        require(_args.from == msg.sender, "Controller: depositor address and msg.sender address mismatch");
+
+        address whitelistModule = AddressBookInterface(addressBook).getWhitelist();
+        WhitelistInterface whitelist = WhitelistInterface(whitelistModule);
+
+        require(
+            whitelist.isWhitelistedCollateral(_args.asset),
+            "Controller: asset is not whitelisted to be used as collateral"
+        );
+
+        vaults[_args.owner][_args.vaultId]._addCollateral(_args.asset, _args.amount, _args.index);
+
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        marginPool.transferToPool(_args.asset, _args.from, _args.amount);
```

Comment:
might need to have a check on args.from == msg.sender depending on what decision we make on if its okay to allow vault operators to take money from the users' addresses directly